src/KubernetesClient/Informers/SharedInformer.cs

```diff
@@ -0,0 +1,184 @@
+                {
+                    _cacheSynchronized.TrySetException(e);
+                    // _cacheSynchronized.OnError(e);
+                    return Observable.Throw<CacheSynchronized<ResourceEvent<TResource>>>(e);
+                })
+                .Finally(() => _cacheSynchronized.TrySetResult(false))
+                // .SubscribeOn(_masterScheduler)
+                .Publish();
+        }
+
+
+        public IObservable<ResourceEvent<TResource>> GetResource(ResourceStreamType type)
+        {
+            var childScheduler = new EventLoopScheduler(); // dedicated thread for the child on which all messages are syncrhonized
+            return Observable.Defer(async () =>
+                {
+                    AddSubscriber();
+                    _logger.LogTrace("Subscriber awaiting cache synchronization before attaching");
+
+                    var isCacheSynchronized = await _cacheSynchronized.Task;
+                    if (!isCacheSynchronized) // really this only happens if the reset is the master completes before first reset, in which case the downstream subscriber gets nothing
+                        return Observable.Empty<ResourceEvent<TResource>>();
+                    // we use lock to pause any processing of the broadcaster while we're attaching to the stream so proper alignment can be made
+
+                    _logger.LogTrace("Subscriber attaching to broadcaster");
+                    return Observable.Create<ResourceEvent<TResource>>(observer =>
+                        {
+                            var broadcasterAttachment = Disposable.Empty;
+                            var cacheSnapshot = _cache.Snapshot();
+                            if (type.HasFlag(ResourceStreamType.List))
+                            {
+                                _logger.LogTrace($"Flushing contents of cache version {cacheSnapshot.Version}");
+                                _cache.Values
+                                    .ToReset(type == ResourceStreamType.ListWatch)
+                                    .ToObservable()
+                                    .Concat(Observable.Never<ResourceEvent<TResource>>())
+                                    .ObserveOn(Scheduler.Immediate)
+                                    .Subscribe(observer);
+                            }
+
+                            if (type.HasFlag(ResourceStreamType.Watch))
+                                broadcasterAttachment = _masterObservable
+                                    // we could be ahead of broadcaster because we initialized from cache which gets updated before the message are sent to broadcaster
+                                    // this logic realigns us at the correct point with the broadcaster
+                                    .Do(x => _logger.LogTrace($"Received from broadcaster {x}"))
+                                    .SkipWhile(x => x.MessageNumber <= cacheSnapshot.Version)
```

Comment:
Can you elaborate why this is `MessageNumber` instead of `CacheVersion`?

Reply:
What ends up happening is when reset completes, cache gets synchronized up to the LAST message is reset block, however the actual messages haven't been "broadcast" yet. The result is that cache version is actually AHEAD of what is being sent out of the broadcaster. When new subscriber joins, they "List" from cache, and their position in the event stream can very well end up ahead of the broadcaster. If they are trying to watch after, they need to skip any messages that were already included in the cache and only start processing them once the version of cache they listed from is equals to what is being sent out of broadcaster. This is the whole reason for CacheSynchronized datastructure as I struggled with this synchronization dilemma for a long time and finally settled on this approach.